build-scripts/config_common.cmake

```diff
@@ -554,3 +554,6 @@ else ()
   # Disable aot intrinsics for interp, fast-jit and llvm-jit
   add_definitions (-DWASM_ENABLE_AOT_INTRINSICS=0)
 endif ()
+if (WAMR_BUILD_MEM_ALLOC_WITH_USAGE EQUAL 1)
```

Comment:
> There is already a cmake variable `WAMR_BUILD_ALLOC_WITH_USER_DATA`, here how about using `WAMR_BUILD_ALLOC_WITH_USAGE` and putting these lines together with `WAMR_BUILD_ALLOC_WITH_USER_DATA`?
> 
I'd like reuse WAMR_BUILD_ALLOC_WITH_USAGE, but that seems will change the API signiture of:
`static void *(*malloc_func)(void *user_data, unsigned int size) = NULL;`
that will impact its user.
BTW, the user defined malloc_func, realloc_func and free_func seems only have a hidden convention with its user, can we make it more abstract and explicitly declared in wamr_export.h? then we can include all of these things in general API.

Reply:
Adding WAMR_BUILD_ALLOC_WITH_USAGE is good (together with WAMR_BUILD_ALLOC_WITH_USER_DATA) to me, but since developer can just use wasm_export.h and libvmlib.a or libiwasm.so for his project, normally we should not add macro control in wasm_export.h, it is not easy to declare function prototypes for malloc_func/free_func, how about we just add comments about the prototypes in:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/42199f163ec02cc899bad828360cb6128b1c4fb3/core/iwasm/include/wasm_export.h#L123-L125

Reply:
And if that, we had better update build_wamr.doc about the two cmake flags, and control the code like:
```C
static void *(*malloc_func)(
#if WASM_MEM_ALLOC_WITH_USAGE
mem_alloc_usage_t usage,
#endif
#if WASM_MEM_ALLOC_WITH_USER_DATA
void *user_data,
#endif
unsigned int size) = NULL;
```

Reply:
what about this in wamr_export.h:
```
typedef enum {
    Alloc_For_Runtime,
    Alloc_For_LinearMemory
} mem_alloc_usage_t;

typedef struct {
    mem_alloc_usage_t usage;
    void *user_data;
} user_data_t;

typedef void* (*malloc_func_t)(const user_data_t *data, uint64_t size);
typedef void* (*realloc_func_t)(const user_data_t *data, void *ptr, uint64_t size);
typedef void* (*free_func_t)(const user_data_t *data, void *ptr);
```
and in wasm_memory.c
```
static void *allocator_user_data = NULL;
static malloc_func_t malloc_func = NULL;
static realloc_func_t realloc_func = NULL;
static free_func_t free_func = NULL;
...
static bool
wasm_memory_init_with_allocator(malloc_func_t malloc_func_ptr, realloc_func_t realloc_func_ptr,
                                free_func_t free_func_ptr)
{
    if (malloc_func_ptr && free_func_ptr && malloc_func_ptr != free_func_ptr) {
        memory_mode = MEMORY_MODE_ALLOCATOR;
        malloc_func = malloc_func_ptr;
        realloc_func = realloc_func_ptr;
        free_func = free_func_ptr;
        return true;
    }
    LOG_ERROR("Init memory with allocator (%p, %p, %p) failed.\n",
              malloc_func_ptr, realloc_func_ptr, free_func_ptr);
    return false;
}
```
then we can merge WASM_MEM_ALLOC_WITH_USAGE&WASM_MEM_ALLOC_WITH_USER_DATA together.

Reply:
My preference is we had better not change the signature of malloc_func/realloc_func functions for general scenarios, they have been used by many developers, changing them will cause many inconvenient.